neqo-common/src/datagram.rs

```diff
@@ -44,17 +41,85 @@ impl Datagram {
     pub fn set_tos(&mut self, tos: IpTos) {
         self.tos = tos;
     }
+
+    #[must_use]
+    pub const fn segment_size(&self) -> usize {
+        self.segment_size
+    }
 }
 
-impl Deref for Datagram {
-    type Target = Vec<u8>;
+impl<D: AsRef<[u8]>> Datagram<D> {
+    pub fn new(
+        src: SocketAddr,
+        dst: SocketAddr,
+        tos: IpTos,
+        d: D,
+        segment_size: Option<usize>,
+    ) -> Self {
+        Self {
+            src,
+            dst,
+            tos,
+            segment_size: segment_size.unwrap_or_else(|| d.as_ref().len()),
+            d,
+        }
+    }
+    pub fn iter_segments(&self) -> impl Iterator<Item = &[u8]> {
+        self.d.as_ref().chunks(self.segment_size)
+    }
+
+    pub fn num_segments(&self) -> usize {
+        self.d.as_ref().len().div_ceil(self.segment_size)
+    }
+}
+
+impl<D: Deref<Target = [u8]>> Deref for Datagram<D> {
+    type Target = [u8];
     #[must_use]
     fn deref(&self) -> &Self::Target {
         &self.d
     }
 }
 
-impl std::fmt::Debug for Datagram {
+impl<'a> From<&'a Datagram> for Datagram<&'a [u8]> {
```

Comment:
Should this be a Borrow implementation instead?

Reply:
I don't think implementing `Borrow` is possible.

Here is the trait definition:

``` rust
pub trait Borrow<Borrowed: ?Sized> {
    fn borrow(&self) -> &Borrowed;
}
```

https://doc.rust-lang.org/alloc/borrow/trait.Borrow.html

The problematic part is the return value, i.e. `&Borrowed`. It would need to return a `&Datagram<&'a [u8]>`. Given that the `Datagram<&'a [u8]>` would need to be instantiated within the `borrow` implementation, the reference would point to a temporary value.

I have created the following simplified playground as a showcase.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=177f0e0f97ebbce7376092c924c72c55